api/service/legacysync/epoch_syncing.go

```diff
@@ -0,0 +1,240 @@
+	})
+}
+
+// isInSync query the remote DNS node for the latest height to check what is the current
+// sync status
+func (ss *EpochSync) isInSync(doubleCheck bool) SyncCheckResult {
+	if ss.syncConfig == nil {
+		return SyncCheckResult{} // If syncConfig is not instantiated, return not in sync
+	}
+	otherHeight1 := getMaxPeerHeight(ss.syncConfig)
+	curBlock := ss.beaconChain.CurrentBlock()
+	epochLastBlock := shard.Schedule.EpochLastBlock(curBlock.Epoch().Uint64())
+	lastHeight := ss.beaconChain.CurrentBlock().NumberU64()
+
+	if !doubleCheck {
+		heightDiff := otherHeight1 - lastHeight
+		if otherHeight1 < lastHeight {
+			heightDiff = 0 //
+		}
+		inSync := epochLastBlock > otherHeight1
+		utils.Logger().Info().
+			Uint64("OtherHeight", otherHeight1).
+			Uint64("lastHeight", lastHeight).
+			Msg("[EPOCHSYNC] Checking sync status")
+		return SyncCheckResult{
+			IsInSync:    inSync,
+			OtherHeight: otherHeight1,
+			HeightDiff:  heightDiff,
+		}
+	}
+	// double check the sync status after 1 second to confirm (avoid false alarm)
+	time.Sleep(1 * time.Second)
+
+	otherHeight2 := getMaxPeerHeight(ss.syncConfig)
+	inSync := epochLastBlock > otherHeight2
+
+	heightDiff := otherHeight2 - lastHeight
+	if otherHeight2 < lastHeight {
+		heightDiff = 0 // overflow
+	}
+
+	return SyncCheckResult{
+		IsInSync:    inSync,
+		OtherHeight: otherHeight2,
+		HeightDiff:  heightDiff,
+	}
+}
+
+// GetActivePeerNumber returns the number of active peers
+func (ss *EpochSync) GetActivePeerNumber() int {
+	if ss.syncConfig == nil {
+		return 0
+	}
+	// len() is atomic; no need to hold mutex.
+	return len(ss.syncConfig.peers)
+}
+
+// SyncLoop will keep syncing with peers until catches up
+func (ss *EpochSync) SyncLoop(bc *core.BlockChain, worker *worker.Worker, isBeacon bool, consensus *consensus.Consensus) {
+	timeout := 2
+	for {
+		<-time.After(time.Duration(timeout) * time.Second)
+		timeout = ss.syncLoop(bc, worker, isBeacon, consensus)
+	}
+}
+
+func (ss *EpochSync) syncLoop(bc *core.BlockChain, worker *worker.Worker, isBeacon bool, _ *consensus.Consensus) (timeout int) {
+	maxHeight := getMaxPeerHeight(ss.syncConfig)
+	for {
+		block := bc.CurrentBlock()
+		height := block.NumberU64()
+		if height >= maxHeight {
+			utils.Logger().Info().
+				Msgf("[EPOCHSYNC] Node is now IN SYNC! (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
+					isBeacon, bc.ShardID(), maxHeight, height)
+			return 60
+		}
+
+		utils.Logger().Info().
+			Msgf("[EPOCHSYNC] Node is OUT OF SYNC (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d, peers count %d)",
+				isBeacon, bc.ShardID(), maxHeight, height, ss.syncConfig.PeersCount())
+
+		var heights []uint64
+		curEpoch := block.Epoch()
+		for i := 0; i < int(SyncLoopBatchSize); i++ {
+			epochLastBlock := shard.Schedule.EpochLastBlock(curEpoch.Uint64())
+			if epochLastBlock > maxHeight {
+				break
+			}
+			if height != epochLastBlock {
+				heights = append(heights, epochLastBlock)
```

Comment:
when i = 0, we are downloading the latest epoch block we already had "block := bc.CurrentBlock()". This is an epoch block and if we get EpochLastBlock of it's epoch. it returns the block itself. So we should skip this first one.